cmd/gpu_nfdhook/main.go

```diff
@@ -0,0 +1,121 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+const (
+	sysfsDirectory    = "/host-sys"
+	devfsDirectory    = "/host-dev"
+	sysfsDrmDirectory = sysfsDirectory + "/class/drm"
+	devfsDriDirectory = devfsDirectory + "/dri"
+	gpuDeviceRE       = `^card[0-9]+$`
+	controlDeviceRE   = `^controlD[0-9]+$`
+	vendorString      = "0x8086"
+
+	namespace = "gpu.intel.com"
+)
+
+type labeler struct {
+	sysfsDir string
+	devfsDir string
+
+	gpuDeviceReg     *regexp.Regexp
+	controlDeviceReg *regexp.Regexp
+}
+
+func newLabeler(sysfsDir, devfsDir string) *labeler {
+	return &labeler{
+		sysfsDir:         sysfsDir,
+		devfsDir:         devfsDir,
+		gpuDeviceReg:     regexp.MustCompile(gpuDeviceRE),
+		controlDeviceReg: regexp.MustCompile(controlDeviceRE),
+	}
+}
+
+func (l *labeler) scan() ([]string, error) {
+	files, err := ioutil.ReadDir(l.sysfsDir)
+	gpuNameList := []string{}
+
+	if err != nil {
+		return gpuNameList, errors.Wrap(err, "Can't read sysfs folder")
+	}
+
+	for _, f := range files {
+		if !l.gpuDeviceReg.MatchString(f.Name()) {
+			klog.V(4).Info("Not compatible device", f.Name())
+			continue
+		}
+
+		dat, err := ioutil.ReadFile(path.Join(l.sysfsDir, f.Name(), "device/vendor"))
+		if err != nil {
+			klog.Warning("Skipping. Can't read vendor file: ", err)
+			continue
+		}
+
+		if strings.TrimSpace(string(dat)) != vendorString {
+			klog.V(4).Info("Non-Intel GPU", f.Name())
+			continue
+		}
+
+		drmFiles, err := ioutil.ReadDir(path.Join(l.sysfsDir, f.Name(), "device/drm"))
+		if err != nil {
+			return gpuNameList, errors.Wrap(err, "Can't read device folder")
+		}
+
+		for _, drmFile := range drmFiles {
+			if l.controlDeviceReg.MatchString(drmFile.Name()) {
+				//Skipping possible drm control node
+				continue
+			}
+			devPath := path.Join(l.devfsDir, drmFile.Name())
+			if _, err := os.Stat(devPath); err != nil {
+				continue
+			}
+
+			gpuNameList = append(gpuNameList, f.Name())
+			break
+		}
+	}
+
+	return gpuNameList, nil
+}
+
+func main() {
+	l := newLabeler(sysfsDrmDirectory, devfsDriDirectory)
+	gpuNames, err := l.scan()
+	if err != nil {
+		klog.Errorf("scan failed:%v", err.Error())
+		return
+	}
+
+	labels, err := l.createLabels(gpuNames)
+
+	if err == nil {
+		for key, val := range labels {
+			fmt.Println(key + "=" + val)
+		}
+	}
+}
```

Comment:
It seems the main function could be refactored to just
```go
func main() {
        l := newLabeler(sysfsDrmDirectory, devfsDriDirectory)
        l.printLabels()
}
```

Then the unit tests could avoid messing with the module's global state.